### PR TITLE
Make declaration from usage

### PR DESCRIPTION
**Pull request by Razzeee**

### CODE REVIEW

**Comment by Razzeee** [comment]
I think this is ready for a first version.

File: src/providers/codeActionProvider.ts [hunk 1]
```diff
@@ -282,4 +287,62 @@ export class CodeActionProvider {
...
+          nodeAtPosition.tree,
+          params.textDocument.uri,
+          elmWorkspace.getImports(),
+        );
+
+        const edit = RefactorEditUtils.createTopLevelFunction(
+          insertLineNumber ?? tree.rootNode.endPosition.row,
+          funcName,
+          typeString,
+          nodeAtPosition.parent?.parent?.parent,
```
Comment (jmbockhorst) [inline src/providers/codeActionProvider.ts:329]: Using `TreeUtils.parentOfType('value_declaration', nodeAtPosition)` might be safer and cleaner.

File: src/providers/codeActionProvider.ts [hunk 2]
```diff
@@ -282,4 +287,62 @@ export class CodeActionProvider {
...
+        );
+
+        if (edit) {
+          codeActions.push({
```
Comment (jmbockhorst) [inline src/providers/codeActionProvider.ts:333]: Should would try to put this code action above ones from elm make? Otherwise it can be easy to miss

**Review by jmbockhorst — APPROVED** [review]
Works good, just a few minor comments. Type inference doesn't work for functions with no args (I knew it wouldn't), so I will add that piece tonight in a separate PR.

File: src/providers/codeActionProvider.ts (hunk 1 above)
Comment (Razzeee) [reply src/providers/codeActionProvider.ts]: Changed

File: src/providers/codeActionProvider.ts (hunk 2 above)
Comment (Razzeee) [reply src/providers/codeActionProvider.ts:335]: I've moved these and the type annotation ones to the top now

**Pull request merged**